x-pack/plugins/security_solution/public/common/components/exceptions/viewer/index.tsx

```diff
@@ -4,96 +4,411 @@
+  CREATE = 'CREATE',
+  EDIT = 'EDIT',
+}
+
+interface ExceptionsViewerProps {
+  ruleId: string;
+  exceptionListsMeta: ExceptionIdentifiers[];
+  availableListTypes: ExceptionListType[];
   commentsAccordionId: string;
-  handleDelete: ({ id }: { id: string }) => void;
-  handleEdit: (item: ExceptionListItemSchema) => void;
+  onAssociateList?: (listId: string) => void;
 }
 
-const ExceptionItemComponent = ({
-  exceptionItem,
+const ExceptionsViewerComponent = ({
+  ruleId,
+  exceptionListsMeta,
+  availableListTypes,
+  onAssociateList,
   commentsAccordionId,
-  handleDelete,
-  handleEdit,
-}: ExceptionItemProps): JSX.Element => {
-  const [entryItems, setEntryItems] = useState<FormattedEntry[]>([]);
-  const [showComments, setShowComments] = useState(false);
+}: ExceptionsViewerProps): JSX.Element => {
+  const { services } = useKibana();
+  const [, dispatchToaster] = useStateToaster();
+  const [initLoading, setInitLoading] = useState(true);
+  const onDispatchToaster = useCallback(
+    ({ title, color, iconType }) => (): void => {
+      dispatchToaster({
+        type: 'addToaster',
+        toast: {
+          id: uuid.v4(),
+          title,
+          color,
+          iconType,
+        },
+      });
+    },
+    [dispatchToaster]
+  );
+  const { deleteExceptionItem } = useApi(services.http);
+  const [
+    {
+      endpointList,
+      detectionsList,
+      exceptions,
+      filterOptions,
+      pagination,
+      loadingItemIds,
+      isModalOpen,
+    },
+    dispatch,
+  ] = useReducer(allExceptionItemsReducer(), initialState);
 
-  useEffect((): void => {
-    const formattedEntries = getFormattedEntries(exceptionItem.entries);
-    setEntryItems(formattedEntries);
-  }, [exceptionItem.entries]);
+  const setExceptions = useCallback(
+    ({
+      lists: newLists,
+      exceptions: newExceptions,
+      pagination: newPagination,
+    }: SetExceptionsProps) => {
+      dispatch({
+        type: 'setExceptions',
+        lists: newLists,
+        exceptions: newExceptions,
+        pagination: newPagination,
+      });
+    },
+    [dispatch]
+  );
+  const [loadingList, , , , fetchList] = useExceptionList({
+    http: services.http,
+    lists: exceptionListsMeta,
+    filterOptions,
+    pagination: {
+      page: pagination.pageIndex + 1,
+      perPage: pagination.pageSize,
+      total: pagination.totalItemCount,
+    },
+    dispatchListsInReducer: setExceptions,
+    onError: onDispatchToaster({
+      color: 'danger',
+      title: i18n.FETCH_LIST_ERROR,
+      iconType: 'alert',
+    }),
+  });
 
-  const onDelete = useCallback((): void => {
-    handleDelete({ id: exceptionItem.id });
-  }, [handleDelete, exceptionItem]);
+  const setIsModalOpen = useCallback(
+    (isOpen: boolean): void => {
+      dispatch({
+        type: 'updateModalOpen',
+        isOpen,
+      });
+    },
+    [dispatch]
+  );
+
+  const onFetchList = useCallback((): void => {
+    if (fetchList != null) {
+      fetchList();
+    }
```

Comment:
Was going to say you could just use optional chaining here, but looks like our version of `prettier` still throws an error when using it with functions... 😔 

![image](https://user-images.githubusercontent.com/2946766/84194836-db39d380-aa5a-11ea-821b-035ba3cffce9.png)

Reply:
Ooo that's kinda cool. It does yell, when I try :(